drivers/gpio/gpio_mcux.c

```diff
@@ -328,10 +325,53 @@ static void gpio_mcux_port_isr(const struct device *dev)
 	/* Clear the gpio interrupts */
 	config->gpio_base->ISFR[0] = int_status;
 #endif  /* !(defined(FSL_FEATURE_PORT_HAS_NO_INTERRUPT) */
+	return int_status;
+}
+
+static void gpio_mcux_port_isr(const struct device *dev)
+{
+	const struct gpio_mcux_config *config = dev->config;
+	struct gpio_mcux_data *data = dev->data;
 
-	gpio_fire_callbacks(&data->callbacks, dev, int_status);
+	gpio_fire_callbacks(&data->callbacks, dev, read_gpio_interrupt_status(config));
 }
 
+#define GPIO_HAS_SHARED_IRQ DT_HAS_COMPAT_STATUS_OKAY(nxp_kinetis_gpio_cluster)
+
+#if GPIO_HAS_SHARED_IRQ
+static void gpio_mcux_shared_cluster_isr(const struct device *ports[])
+{
+	const struct device **current_port = &ports[0];
+
+	while (*current_port != NULL) {
+		gpio_mcux_port_isr(*current_port);
+		current_port++;
+	}
+}
+
+#define CLUSTER_ARRAY_ELEMENT(node_id) DEVICE_DT_GET(node_id),
+
+#define GPIO_MCUX_CLUSTER_INIT(node_id)								\
+	const struct device *SharedArray##node_id[DT_CHILD_NUM_STATUS_OKAY(node_id) + 1] =	\
+		{DT_FOREACH_CHILD_STATUS_OKAY(node_id, CLUSTER_ARRAY_ELEMENT) NULL};	\
+												\
+	static int gpio_mcux_shared_interrupt_init##node_id(void)	\
+	{								\
+		do {							\
```

Comment:
you can remove the do while false since this code is already wrapped by the function definition